Adding YouTube video support

farhan-arshad-dev opened the pull request:
### Description

[LEARNER-7469](https://openedx.atlassian.net/browse/LEARNER-7469)
- Adding support to play youtube videos inside the app, avoiding the current redirection to the youtube app. These changes support transcripts while maintaining default video behavior.
- Add utility method to get the Youtube App version that is used to check either the supported version of YouTube app is installed in the device or not, because YouTube in-app player SDK needs  YouTube app to be installed, but the feature has been tested over the **v10.18.55** and this has not worked with message "**An error occurred while initializing the YouTube player.**", so the target version has been set in the minor **v12.01.54** where it works.
- Make the code common to download, maintain and display the transcripts for the video player screens by refactoring the 'TranscriptManager.java and 'TranscriptDownloader.java'.
- Remove the wait logic to fetch the download transcripts.
 
**Ref:** 
- https://stackoverflow.com/a/37553255
- https://developers.google.com/youtube/android/player/.

**Screenshots:**

| Youtube Videos Listing  | YouTube Player |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/43750646/68470678-5b3a5980-023e-11ea-8335-5a202b983b34.png"/>  | <img src="https://user-images.githubusercontent.com/43750646/68470680-5bd2f000-023e-11ea-9ab0-5cff017bf183.png" /> |

**Notes**
Feature is behind a flag
```
YOUTUBE_IN_APP_PLAYER:
    ENABLED: True
    YOUTUBE_PLAYER_API_KEY: 'YOUR_API_KEY'  
```
Open a youtube video available in the course

**Acceptance Criteria**:
- YouTube video content should be played in the app.
- Transcripts should be visible with the video content (YouTube and other encoded) if available.
- The app should not crash especially for the player screens.

miankhalid commented:
This feature introduces a feature-flag in our configurations repo. Please do the following:
1. Create a ticket for sandbox configuration.
2. Update the configs.
3. Update the [App Configuration Flags](https://openedx.atlassian.net/wiki/spaces/LEARNER/pages/48792067/App+Configuration+Flags) wiki.

farhan-arshad-dev commented:
> This feature introduces a feature-flag in our configurations repo. Please do the following:
> 
> 1. Create a ticket for sandbox configuration.
> 2. Update the configs.
> 3. Update the [App Configuration Flags](https://openedx.atlassian.net/wiki/spaces/LEARNER/pages/48792067/App+Configuration+Flags) wiki.

Done.

farhan commented:
We should provide the casting options for Youtube videos.

farhan-arshad-dev commented:
> We should provide the casting options for Youtube videos.

@farhan 
We try to provide the casting option for YouTube videos, but google cast only received the streamable link, where Youtube video link is the Http page address, and there is no directly streamable link.
So after the discussion with the IOS team, we decide that we cover this case in the scope of another [story](https://openedx.atlassian.net/browse/LEARNER-7477).

farhan commented:
**Issues:**

1. More videos button should be disabled because it's redirecting the user to youtube videos that are not related to our app and which end up in redirecting the user to the system youtube app.

2. Sometimes content below the video disappears.

<img src="https://user-images.githubusercontent.com/25842457/69533612-eb480500-0f99-11ea-9f05-f4c8b3b30eea.png" width="200">

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/Config.java:
```diff
@@ -61,6 +61,7 @@
     private static final String PUSH_NOTIFICATIONS_FLAG = "PUSH_NOTIFICATIONS";
     private static final String WHITE_LIST_OF_DOMAINS = "WHITE_LIST_OF_DOMAINS";
     private static final String API_URL_VERSION = "API_URL_VERSION";
+    private static final String YOUTUBE_VIDEO = "YOUTUBE_VIDEO";
```
@saeedbashir thoughts?

farhan-arshad-dev replied:
I think the iOS app is also using the same config name. am I right? @saeedbashir

saeedbashir replied:
The configurations are not for the player but for the youtube videos. `YOUTUBE_VIDEO` seems a better name.

miankhalid replied:
Since we don't have a parent level config dedicated for video handling, having the config's name as `YOUTUBE_VIDEO` doesn't make any sense.

I would recommend we go with the following config:

```
YOUTUBE_IN_APP_PLAYER:
   ENABLED: true
   API_KEY: 'api_key_goes_here'
```

farhan-arshad-dev replied:
done.

farhan-arshad-dev commented:
> 1. More videos button should be disabled because it's redirecting the user to youtube videos that are not related to our app and it ends up redirecting the user to the system youtube app.
> 
> <img alt="" width="200" src="https://user-images.githubusercontent.com/25842457/69533612-eb480500-0f99-11ea-9f05-f4c8b3b30eea.png">

No way found to remove the more videos button from the player, either in the documentation.

farhan-arshad-dev commented:
> 1. Very first time when you open some youtube app, transcripts don't load.
> 
> <img alt="" width="200" src="https://user-images.githubusercontent.com/25842457/69784280-3c910800-11d7-11ea-9211-eab1928b76ba.png">
> 
> 1. In the follow up of the second step, enter into full-screen mode and then come back to portrait mode, you will see the bottom section will get disappeared.

Fixed

> 1. In the follow up of the second step, enter into full-screen mode and then come back to portrait mode, you will see the bottom section will get disappeared.
> 
> <img alt="" width="200" src="https://user-images.githubusercontent.com/25842457/69784279-3c910800-11d7-11ea-9d13-8ba52d2f310c.png">

Fixed

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -638,7 +638,7 @@ public void onPlaybackPaused() {
                 double current_time = player.getCurrentPosition()/AppConstants.MILLISECONDS_PER_SECOND ;
                 environment.getAnalyticsRegistry().trackVideoPause(videoEntry.videoId, current_time,
-                        videoEntry.eid, videoEntry.lmsUrl);
+                        videoEntry.eid, videoEntry.lmsUrl, "");
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/player/PlayerFragment.java:
```diff
@@ -910,7 +910,7 @@ public void onFullScreen(boolean isFullScreen) {
             environment.getAnalyticsRegistry().trackVideoOrientation(videoEntry.videoId,
                     player.getCurrentPosition() / AppConstants.MILLISECONDS_PER_SECOND,
-                    isFullScreen, videoEntry.eid, videoEntry.lmsUrl);
+                    isFullScreen, videoEntry.eid, videoEntry.lmsUrl, "");
```
done.

farhan-arshad-dev commented on OpenEdXMobile/res/layout/fragment_course_unit_video.xml:
```diff
@@ -63,12 +63,19 @@
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:visibility="gone" />
+    </FrameLayout>
```
Cast Mini controller visibility is controlled by the SDK, and in case of YouTube in-app player screen app needs to play the video within the app, either Connected with the cast device or not. 

So we need to hide the mini controller when another video is casting on the remote device.

farhan-arshad-dev replied:
Application is crashing while placing the `**MiniControllerFragment**` in the container in Runtime, and there is no other way mention in the Chrome cast doc.

farhan-arshad-dev commented on OpenEdXMobile/res/values-fr/strings.xml:
```diff
@@ -351,6 +351,10 @@
+  <!-- Label showing that you need to update youtube in order to watch the video -->
+  <string name="assessment_needed_updating_youtube">Cette vidéo nécessite la mise à jour de l\'application YouTube.</string>
+  <!-- Label for button to update youtube -->
+  <string name="assessment_update_youtube">Mettre à jour YouTube</string>
```
Done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/analytics/Analytics.java:
```diff
@@ -598,6 +600,8 @@ void trackCastDeviceConnectionChanged(@NonNull String eventName, @NonNull String
         // Casting Device Types
         String GOOGLE_CAST = "google_cast";
+        // YouTube Player Type
+        String YOUTUBE = "youtube";
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/Config.java:
```diff
@@ -592,6 +593,22 @@ public String getBlocksApiVersion() {
+        @SerializedName("ENABLED")
+        private boolean enabled;
+
+        @SerializedName("YOUTUBE_API_KEY")
```
Renamed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/Config.java:
```diff
@@ -592,6 +593,22 @@ public String getBlocksApiVersion() {
         }
     }
 
+    public static class EmbeddedYoutubeConfig {
```
Renamed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/Config.java:
```diff
@@ -592,6 +593,22 @@ public String getBlocksApiVersion() {
+            return enabled;
+        }
+
+        public String getYoutubeApiKey() {
```
Renamed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/AppConstants.java:
```diff
@@ -24,6 +24,10 @@
     public static final String VIDEO_FORMAT_MP4 = ".mp4";
     public static final String APP_URI_SCHEME = "org.edx.mobile.innerlinks://";
 
+    public static final String YOUTUBE_APP_URI_PLAY_STORE = "market://details?id=com.google.android.youtube";
```
removed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/VideoUtil.java:
```diff
@@ -152,4 +155,28 @@ public static String getVideoPath(Context context, DownloadEntry video) {
+        } catch (PackageManager.NameNotFoundException e) {
+            return false;
+        }
+    }
```
Done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitOnlyOnYoutubeFragment.java:
```diff
@@ -31,16 +31,22 @@ public void onCreate(Bundle savedInstanceState) {
-                i.setData(Uri.parse(((VideoBlockModel) unit).getData().encodedVideos.youtube.url));
-                startActivity(i);
-            }
+        final View courseUnitOnlyOnYoutubeView = inflater.inflate(R.layout.fragment_course_unit_only_on_youtube, container, false);
```
Renamed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/adapters/CourseUnitPagerAdapter.java:
```diff
@@ -74,15 +77,18 @@ public Fragment getItem(int pos) {
             videoBlockModel.setVideoThumbnail(courseData.getCourse().getCourse_image());
-            unitFragment = CourseUnitVideoFragment.newInstance(videoBlockModel, (pos < unitList.size()), (pos > 0));
-        } else if (minifiedUnit instanceof VideoBlockModel && ((VideoBlockModel) minifiedUnit).getData().encodedVideos.getYoutubeVideoInfo() != null) {
+            unitFragment = CourseUnitAndroidVideoPlayerFragment.newInstance((VideoBlockModel) minifiedUnit, (pos < unitList.size()), (pos > 0));
```
`CourseUnitVideoFragment` is renamed to `CourseUnitAndroidVideoPlayerFragment`

farhan-arshad-dev replied:
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitOnlyOnYoutubeFragment.java:
```diff
@@ -31,16 +31,22 @@ public void onCreate(Bundle savedInstanceState) {
+            ((TextView) courseUnitOnlyOnYoutubeView.findViewById(R.id.only_youtube_available_message)).setText(R.string.assessment_needed_updating_youtube);
+            courseUnitOnlyOnYoutubeView.findViewById(R.id.update_youtube_button).setVisibility(View.VISIBLE);
+            courseUnitOnlyOnYoutubeView.findViewById(R.id.update_youtube_button).setOnClickListener(v -> {
+                final Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(AppConstants.BROWSER_PLAYSTORE_YOUTUBE_URI));
```
```                
final Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(AppConstants.BROWSER_PLAYSTORE_YOUTUBE_URI));
startActivity(intent);
```
this code is automatically redirected to the PlayStore app if installed, otherwise open the link in browser, so there is not need other line of code.

farhan-arshad-dev replied:
> We should also make this kind of functions(in which intent is Intent.ACTION_VIEW) a utility function.

Use an already existing method of `BrowserUtil`

farhan-arshad-dev commented:
@farhan Ready for another pass.

farhan commented:
When `YOUTUBE_VIDEO` tag is present in configuration but `YOUTUBE_PLAYER_API_KEY` is missing in configuration, the app should write error logs.

farhan commented:
> 3. In the follow up of the second step, enter into full-screen mode and then come back to portrait mode, you will see the bottom section will get disappeared.
> <img src="https://user-images.githubusercontent.com/25842457/69784279-3c910800-11d7-11ea-9d13-8ba52d2f310c.png" width="200">
> 

This issue still exists

farhan commented:
Youtube player should be reinitialized if required on internet connectivity.

farhan commented:
Why this screen has changed ?

| Old | New |
| --- | --- |
| <img src="https://user-images.githubusercontent.com/25842457/70502490-b06dd180-1b42-11ea-9d4f-699cd6b9e9e9.png" width="200"> | <img src="https://user-images.githubusercontent.com/25842457/70502494-b2379500-1b42-11ea-9847-5afeecd72b5d.png" width="200"> |

farhan commented:
Release blockers reported by the Testing team should be fixed within the scope of this PR.

farhan-arshad-dev commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -425,6 +425,11 @@
 
+    <!-- Label showing that you need to update youtube in order to watch the video -->
+    <string name="assessment_needed_updating_youtube" tools:ignore="MissingTranslation">This video requires that you update the YouTube app.</string>
+    <!-- Label for button to update youtube -->
```
IMO `Label for button to update youtube app`. is more meaning full.

farhan-arshad-dev commented on OpenEdXMobile/src/test/java/org/edx/mobile/view/CourseUnitNavigationActivityTest.java:
```diff
@@ -407,7 +406,11 @@ public void testUnitFragmentCreation() {
         when(youtubeVideosModel.getData()).thenReturn(videoData2);
         unitList.add(youtubeVideosModel);
-        classesList.add(CourseUnitOnlyOnYoutubeFragment.class);
+        if (config.getYoutubeInAppPlayerConfig().isYoutubeEnabled()) {
```
`[LEARNER-6713](https://openedx.atlassian.net/browse/LEARNER-6713)` this story is not merged yet.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/AppConstants.java:
```diff
@@ -24,6 +24,9 @@
     public static final String VIDEO_FORMAT_MP4 = ".mp4";
     public static final String APP_URI_SCHEME = "org.edx.mobile.innerlinks://";
 
+    public static final String BROWSER_PLAYSTORE_YOUTUBE_URI = "https://play.google.com/store/apps/details?id=com.google.android.youtube";
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/VideoUtil.java:
```diff
@@ -152,4 +155,33 @@ public static String getVideoPath(Context context, DownloadEntry video) {
+    public static boolean isYoutubeAPISupported(Context context) {
+        try {
+            final PackageInfo packageInfo = context.getPackageManager().getPackageInfo(YOUTUBE_PACKAGE_NAME, 0);
+            /*
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitAndroidVideoPlayerFragment.java:
```diff
@@ -0,0 +1,347 @@
+
+import subtitleFile.Caption;
+
+public class CourseUnitAndroidVideoPlayerFragment extends CourseUnitVideoFragment {
```
renamed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -49,57 +33,43 @@
+public abstract class CourseUnitVideoFragment extends CourseUnitFragment
+        implements IPlayerEventCallback, TranscriptListener {
+
+    protected abstract void seekToCaption(Caption caption);
```
Added

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -693,4 +358,8 @@ public void updateBottomSectionVisibility(int visibility) {
         }
     }
+
+    public boolean showMiniController() {
```
renamed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -632,24 +299,9 @@ public void updateSelection(final int subtitleIndex) {
     }
 
-    private void initTranscriptListView() {
+    protected void initTranscriptListView() {
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -49,57 +33,43 @@
 
-public class CourseUnitVideoFragment extends CourseUnitFragment
-    implements IPlayerEventCallback, TranscriptListener {
+public abstract class CourseUnitVideoFragment extends CourseUnitFragment
```
renamed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitAndroidVideoPlayerFragment.java:
```diff
@@ -0,0 +1,347 @@
+
+import subtitleFile.Caption;
+
+public class CourseUnitAndroidVideoPlayerFragment extends CourseUnitVideoFragment {
```
added.

farhan-arshad-dev commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -425,6 +425,11 @@
     <!-- Label for button to open on youtube -->
     <string name="assessment_open_on_youtube">View video on YouTube</string>
 
```
removed.

farhan-arshad-dev commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -425,6 +425,11 @@
     <!-- Label for button to open on youtube -->
     <string name="assessment_open_on_youtube">View video on YouTube</string>
 
+    <!-- Label showing that you need to update youtube in order to watch the video -->
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitOnlyOnYoutubeFragment.java:
```diff
@@ -31,16 +30,19 @@ public void onCreate(Bundle savedInstanceState) {
+            ((TextView) view.findViewById(R.id.only_youtube_available_message)).setText(R.string.assessment_needed_updating_youtube);
+            view.findViewById(R.id.update_youtube_button).setVisibility(View.VISIBLE);
+            view.findViewById(R.id.update_youtube_button).setOnClickListener(v -> {
+                BrowserUtil.open(getActivity(), AppConstants.BROWSER_PLAYSTORE_YOUTUBE_URI);
```
We are just passing some HTTP link that opens the play store app to update the YouTube if installed, otherwise, open the web browser.

IMO, We can just use `BrowserUtil.open` method to do that.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/Router.java:
```diff
@@ -55,6 +55,7 @@
     public static final String EXTRA_DISCUSSION_THREAD_ID = "discussion_thread_id";
     public static final String EXTRA_DISCUSSION_COMMENT = "discussion_comment";
     public static final String EXTRA_DISCUSSION_TOPIC_ID = "discussion_topic_id";
+    public static final String EXTRA_FROM_MY_VIDEOS = "FromMyVideos";
```
removed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitFragment.java:
```diff
@@ -33,4 +33,8 @@ public void onCreate(Bundle savedInstanceState) {
     public void setHasComponentCallback(HasComponent callback) {
         hasComponentCallback = callback;
     }
+
```
added.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitYoutubeVideoFragment.java:
```diff
@@ -0,0 +1,414 @@
+import subtitleFile.FormatSRT;
+import subtitleFile.TimedTextObject;
+
+public class CourseUnitYoutubeVideoFragment extends CourseUnitVideoFragment implements YouTubePlayer.OnInitializedListener {
```
renamed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitYoutubeVideoFragment.java:
```diff
@@ -0,0 +1,414 @@
+    /**
+     * This runnable is used the fetch the Subtitle in TimedTextObject
+     */
+    private Runnable subtitleFetchProcessor = new Runnable() {
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/test/java/org/edx/mobile/view/CourseUnitNavigationActivityTest.java:
```diff
@@ -407,7 +406,11 @@ public void testUnitFragmentCreation() {
+        if (config.getYoutubeInAppPlayerConfig().isYoutubeEnabled()) {
+            classesList.add(CourseUnitYoutubeVideoFragment.class);
+        } else {
+            classesList.add(CourseUnitOnlyOnYoutubeFragment.class);
```
Updated.

farhan-arshad-dev commented:
@farhan Ready for another review.

farhan commented:
🐛**Bug:**
In this code review pass, Youtube player seekbar is not resuming from the last left position of a video.

farhan-arshad-dev commented:
> bugBug:
> In this code review pass, Youtube player seekbar is not resuming from the last left position of a video.

Fixed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/BaseCourseUnitVideoFragment.java:
```diff
@@ -0,0 +1,496 @@
+                             Bundle savedInstanceState) {
+        View view = inflater.inflate(R.layout.fragment_course_unit_video, container, false);
+        messageContainer = view.findViewById(R.id.message_container);
+        transcriptListView = (ListView) view.findViewById(R.id.transcript_listview);
```
removed.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/BaseCourseUnitVideoFragment.java:
```diff
@@ -0,0 +1,496 @@
+import subtitleFile.Caption;
+import subtitleFile.TimedTextObject;
+
+/**
```
updated.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/BaseCourseUnitVideoFragment.java:
```diff
@@ -0,0 +1,496 @@
+        transcriptAdapter = new TranscriptAdapter(getContext(), environment);
+        transcriptListView.setAdapter(transcriptAdapter);
+
+        transcriptListView.setOnTouchListener((v, event) -> {
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitYoutubePlayerFragment.java:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    @Override
+    public void onInitializationFailure(YouTubePlayer.Provider provider,
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/TranscriptDownloader.java:
```diff
@@ -12,20 +13,24 @@
+/**
+ * Class to download the transcript in background thread and post the result on UI thread
+ * that helps to update the UI component
+ */
```
done.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/util/LocaleUtils.java:
```diff
@@ -133,4 +139,55 @@ private static void sortBasedOnLocale(List<FormOption> options) {
         return null;
     }
+
+    /**
```
updated.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitNavigationActivity.java:
```diff
@@ -70,8 +70,25 @@ protected void onCreate(Bundle savedInstanceState) {
+
             @Override
             public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
+                /*
```
updated.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitYoutubePlayerFragment.java:
```diff
@@ -0,0 +1,308 @@
+     * isInForeground is set on false when the app comes to background from foreground
+     * so this allow to play a video when the app comes to foreground from background
+     */
+    private boolean isInForeground = true;
```
updated.

farhan-arshad-dev commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitYoutubePlayerFragment.java:
```diff
@@ -0,0 +1,308 @@
+             */
+            if (attempts <= 3) {
+                releaseYoutubePlayer();
+                initializeYoutubePlayer();
```
done.

farhan-arshad-dev commented:
@farhan Ready for another pass.

farhan commented:
Youtube icon not appearing when youtube configs are present in the config file.
<img src="https://user-images.githubusercontent.com/25842457/72048197-0f5b6d80-32de-11ea-962d-29918fc6d0bd.png" width="200">
Course: "Software Testing Fundamentals" -> Videos -> Road Map

farhan commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/CourseUnitYoutubePlayerFragment.java:
```diff
@@ -165,6 +165,22 @@ public void onInitializationSuccess(Provider provider,
+    }
+
+    private void redirectToYoutubeDialog() {
+        if (getActivity() != null && !getActivity().isDestroyed()) {
```
Shouldn't we release the player before the showing the error dialog ?

farhan commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -420,6 +420,10 @@
+    <!-- Label showing only when youtube in-app player produce error -->
+    <string name="assessment_unable_to_play_video" tools:ignore="MissingTranslation">Player Error</string>
+    <!-- Error message showing only when youtube in-app player produce error -->
+    <string name="assessment_unable_to_play_video_message" tools:ignore="MissingTranslation">Unable to play this video</string>
```
We can be more elaborative in the message like
`Unable to play this video at the moment.\nYou can view it on youtube app or watch it later.`
And show 'Watch later' button instead of 'Cancel'
Get it approved from Greg or Khalid.

miankhalid replied:
@shamck which error message sounds better to you when the in-app YouTube player is unable to play a video for whatever reason:

Option 1: Unable to play this video
Option 2: Unable to play this video at the moment. You can view it on the YouTube app or watch it later.

> And show 'Watch later' button instead of 'Cancel'

Watch later gives a different context in terms of how YouTube uses it (like queuing a video for the future). I would suggest we go with `OK` as the text for the button to dismiss the error dialog. Thoughts @shamck?

shamck replied:
With regards to the error message, how about this slight variant of Option 2:
"Unable to play this video. You can view it on the YouTube app or try again later." 

Yes, I agree, let's go with "OK" as the button text

miankhalid replied:
Sounds good to me @shamck. Thanks!

farhan commented:
@farhan-arshad-dev I am done with the pass

farhan commented:
@farhan-arshad-dev LGTM now 👍 
Well work done in the restructuring of code 🌟 
Squash the commits and merge it.
